### PR TITLE
Translate sample transmission calculator into workbench

### LINKED ISSUE #28691: Rewrite the Sample Transmission Calculator for the workbench
This is a pretty simple GUI, rewrite it in python following MVP with tests
image

### PR DESCRIPTION
**Pull request by Harrietbrown**
This PR adds the sample transmission calculator, originally only available in mantidplot, to workbench.

Report to: Stephen King

**To test:**

1. Build mantid and open `Interfaces->General->Sample Transmission Calculator`
2. Enter in some values for histogram bins and chemical cormula.
3. Click Calulate and check that the plot is updated with the correct output.
4. Try with invalid inputs and check that the validator allerts you that they are incorect.
1. Try in both single and multiple types for histogram and number and mass density.

Fixes #28691 

---

#### Reviewer ####

Please comment on the following ([full description](http://developer.mantidproject.org/ReviewingAPullRequest.html)):

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there is GUI work does it follow the [GUI standards](http://developer.mantidproject.org/Standards/GUIStandards.html)?
- If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

### CODE REVIEW

File: scripts/SampleTransmissionCalculator/stc_presenter.py [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+                self.view.set_error_indicator(key)
+                warning += validations[key] + ' '
+            self.view.set_validation_label(warning)
+
+    def help_window(self):
+        gui_name = 'Sample Transmission Calculator'
+        collection_file = os.path.join(mantid._bindir, '../docs/qthelp/MantidProject.qhc')
+        version = ".".join(mantid.__version__.split(".")[:2])
+        qt_url = 'qthelp://org.sphinx.mantidproject.' + version + '/doc/interfaces/Sample Transmission Calculator.html'
+        external_url = 'http://docs.mantidproject.org/nightly/interfaces/Sample Transmission Calculator.html'
```
Comment (StephenSmith25) [inline scripts/SampleTransmissionCalculator/stc_presenter.py:48]: This currently isn’t opening the Qt help pages for me. I'd be inclined to change this to use InterfaceManger directly, so something like
```
gui_name = 'Sample Transmission Calculator'
InterfaceManager().showCustomInterfaceHelp(gui_name)
```
should work.

**Review by StephenSmith25 — CHANGES_REQUESTED** [review]
Looks good, function is very similar to the MantidPlot interface, just some minor comments.

Things I found from testing:
- This creates a new interface in MantidPlot as well, so there are two Sample transmission entries in MantidPlot now, one of which doesn't work.
- Sometimes the numbers on the plot axes are not formatted well, e.g try 
`Chemical: C - Low: 0, width: 1, High:5` 
For me the y axis looks a bit off - maybe need to fix the precision of the numbers.
- If you type something wrong in the chemical box it brings up a warning and sends the interface to the back, can be slightly annoying. 

- For me the help button doesn’t bring up the Qt help docs and instead opens a webpage.

File: scripts/SampleTransmissionCalculator/stc_view.py [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+    def plot(self, x, y):
+        self.axes.cla()
+        self.axes.plot(x, y)
+        self.plot_frame.draw()
+
+    def set_validation_label(self, warning_text=''):
+        self.validation_label.setText(warning_text)
+
+    def clear_error_indicator(self):
+        getattr(self, 'histogram_err').setText('')
```
[the remark this answers is not in the document]
Comment (Harrietbrown) [reply scripts/SampleTransmissionCalculator/stc_view.py:70]: It was a holdover from using it in the `set_error_indicator` function, you're right it is fine just calling it on the attribute directly.

**Comment by Harrietbrown** [comment]
- I have stoped the new gui from appearing in mantidplot.
- I don't have a validator for the chemical formula but I've had it attempt to run the algorithm and catch a runtime error.
- The issue with the plot problem is a result of have such a small difference between the transmission I'm not sure if the input values you have used are realistic enough to come up often.
- The help button should be fixed now.

**Review by StephenSmith25 — CHANGES_REQUESTED** [review]
I'm happy with this, thanks for all the changes, they've addressed all the issues I had. Just two small things left:

File: scripts/test/sample_transmission_calculator/CMakeLists.txt [hunk 3]
```diff
@@ -0,0 +1,11 @@
+# Tests for Sample Transmission Calculator
+
+set(TEST_PY_FILES
+    stc_model_test.py
+    stc_presenter_test.py
+    stc_view_test.py
+    )
+
+set(PYUNITTEST_QT_API pyqt5)
+pyunittest_add_test(${CMAKE_CURRENT_SOURCE_DIR} python.scripts
```
Comment (martyngigg) [inline scripts/test/sample_transmission_calculator/CMakeLists.txt:10]: Minor comment: Can we put `python.scripts.sampletransmissioncalculator` here? This string is used as the prefix in the test name that ctest uses. It makes running things with `ctest R` easier if they have a more identifiable prefix.

File: docs/source/release/v5.1.0/mantidworkbench.rst [hunk 4]
```diff
@@ -53,6 +54,7 @@ Improvements
 - Variables assigned in python scripts are now cleared when a script is run in its entirety.
 - The colorbar on colorfill plots is now labelled.
 - User data directories are no longer checked at startup, reducing launch times with slow network drives.
+- The Sample Transmission Calculator is now implemented in workbench.
```
Comment (StephenSmith25) [inline docs/source/release/v5.1.0/mantidworkbench.rst:57]: This line appears twice.

File: scripts/test/sample_transmission_calculator/stc_view_test.py [hunk 5]
```diff
@@ -0,0 +1,100 @@
...
+            'single_high': 10.0,
+            'multiple_bin': '0.1,0.1,20.0',
+            'chemical_formula': 'C',
+            'density_type': 'Number Density',
+            'density': 0.2,
+            'thickness': 0.2,
+        }
+        self.assertEqual(input_dict, input_dict_expected)
+
+    @unittest.skipIf(not hasattr(QHBoxLayout, 'replaceWidget'), 'Invalid qt version')
```
Comment (StephenSmith25) [inline scripts/test/sample_transmission_calculator/stc_view_test.py:45]: Given these are running in QT5 now we should be able to remove these `unittest.skipIf` statements now.

**Review by StephenSmith25 — COMMENTED** [review]
Looking good, just two minor comments and its good to go for me.

**Review by StephenSmith25 — COMMENTED** [review]
Good work, code looks good 👍 .

I just went for a last test and it appears to me that the multiple options doesn't behave exactly the same as the option in MantidPlot. For instance, the one in workbench will only accept three values, whereas MantidPlot accepts any odd number of values. Is this intentional?

**Comment by Harrietbrown** [comment]
I'm not sure what you mean. Do you mean the input wavelength range? because trying on plot it is only accepting when 3 values are given

**Comment by StephenSmith25** [comment]
> I'm not sure what you mean. Do you mean the input wavelength range? because trying on plot it is only accepting when 3 values are given

Maybe I was wrong about this then, if so I'll approve. But this was changing the Input wavelength range type to `multiple`. In plot it was accepting any odd number of values, whereas in Workbench it was only accepting three values.

**Comment by Harrietbrown** [comment]
Ok that change should handle it.

**Review by StephenSmith25 — APPROVED** [review]
Looks good to go 👍

**Pull request merged**